R/backend-snowflake.R

```diff
@@ -18,13 +18,193 @@ NULL
+      # str_detect-like behavior
+      str_detect = function(string, pattern, negate = FALSE) {
+        if (isTRUE(negate)) {
+          sql_expr(!(((!!string)) %REGEXP% (".*" || (!!pattern) || ".*")))
+        } else {
+          sql_expr(((!!string)) %REGEXP% (".*" || (!!pattern) || ".*"))
+        }
+      },
+      # On Snowflake, REGEXP_REPLACE is used like this:
+      # REGEXP_REPLACE( <subject> , <pattern> [ , <replacement> ,
+      #                <position> , <occurrence> , <parameters> ] )
+      # so we must set <occurrence> to 1 if not replacing all.  See:
+      # https://docs.snowflake.com/en/sql-reference/functions/regexp_replace.html
+      # Also, Snowflake needs backslashes escaped, so we must increase the
+      # level of escaping by 1
+      str_replace = function(string, pattern, replacement) {
+        pattern <- gsub("\\\\", "\\\\\\\\", pattern)
+        sql_expr(regexp_replace(!!string, !!pattern, !!replacement, 1, 1))
+      },
+      str_replace_all = function(string, pattern, replacement) {
+        pattern <- gsub("\\\\", "\\\\\\\\", pattern)
+        sql_expr(regexp_replace(!!string, !!pattern, !!replacement))
+      },
+      str_remove = function(string, pattern) {
+        pattern <- gsub("\\\\", "\\\\\\\\", pattern)
+        sql_expr(regexp_replace(!!string, !!pattern, "", 1, 1))
+      },
+      str_remove_all = function(string, pattern) {
+        pattern <- gsub("\\\\", "\\\\\\\\", pattern)
+        sql_expr(regexp_replace(!!string, !!pattern))
+      },
+      str_trim = function(string) {
+        sql_expr(trim(!!string))
+      },
+      str_squish = function(string) {
+        sql_expr(regexp_replace(trim(!!string), "\\\\s+", " "))
+      },
+
+
+      # lubridate functions
+      # https://docs.snowflake.com/en/sql-reference/functions-date-time.html
+      day = function(x) {
+        sql_expr(EXTRACT(DAY %FROM% !!x))
+      },
+      mday = function(x) {
+        sql_expr(EXTRACT(DAY %FROM% !!x))
+      },
+      wday = function(x, label = FALSE, abbr = TRUE, week_start = NULL) {
+        if (!label) {
+          week_start <- week_start %||% getOption("lubridate.week.start", 7)
+          offset <- as.integer(7 - week_start)
+          sql_expr(EXTRACT("dayofweek", DATE(!!x) + !!offset) + 1)
+        } else if (label && !abbr) {
+          sql_expr(
+            DECODE(
+              EXTRACT("dayofweek", !!x),
+              1, "Monday",
+              2, "Tuesday",
+              3, "Wednesday",
+              4, "Thursday",
+              5, "Friday",
+              6, "Saturday",
+              0, "Sunday"
+            )
+          )
+        } else if (label && abbr) {
+          sql_expr(DAYNAME(!!x))
+        } else {
+          abort("Unrecognized arguments to `wday`")
+        }
+      },
+      yday = function(x) sql_expr(EXTRACT("dayofyear", !!x)),
+      week = function(x) {
+        sql_expr(FLOOR((EXTRACT("dayofyear", !!x) - 1L) / 7L) + 1L)
```

Comment:
Are you sure that's the correct calculation? I think it might be too simple.

Reply:
From the `lubridate` docs,
> `week()` returns the number of complete seven day periods that have occurred between the date and January 1st, plus one.
>
> `isoweek()` returns the week as it would appear in the ISO 8601 system, which uses a reoccurring leap week.

I think it _should_ be this simple, and I tested it against both the `lubridate` and postgres-dbplyr implementations for a bunch of dates.

Reply:
Ok, great.